tests/5.2/declare_mapper/test_declare_mapper_present.c

```diff
@@ -0,0 +1,67 @@
+//===---- test_declare_mapper_present.c -------------------------------------===//
+// 
+// OpenMP API Version 5.2 
+//
+// This example has been adapted from the 5.2 OpenMP Examples document:
+// "declare mapper Directive", and "OpenMP Directive Syntax -> complex iterator"
+// The declare mapper directive will be used to automatically map variables
+// according to its prescription:  full structure, plus the dynamic storage of the
+// data element. These variable should be present on the device.
+//
+//===-------------------------------------------------------------------------------===//
+
+#include <omp.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "ompvv.h"
+
+#define N 1000
+
+typedef struct myvec{
+    size_t len;
+    int *data;
+} myvec_t;
+
+
+
+void init( myvec_t *s )
+{ 
+  for(int i = 0; i < s->len; i++)
+    s->data[i] = i; 
+}
+
+
+int test_declare_mapper_present() { 
+
+  OMPVV_INFOMSG("test_declare_mapper_present");
+  int errors = 0;
+
+  myvec_t s;
+
+  s.data = (int *)calloc(N,sizeof(int));
+  s.len  = N;
+  #pragma omp target enter data map(to: s, s.data[0:s.len])
+
+  #pragma omp declare mapper(default:myvec_t v) map(present, from: v, v.data[0:v.len]) 
+
+  #pragma omp target map(s)
+  {
+    init(&s);
+  }
+
+  for (int i = 0; i < N; ++i) {
+    OMPVV_TEST_AND_SET(errors, s.data[i] != i);
```

Comment:
I disagree that the present modifier would prohibit the from map-type from taking effect, as the struct is present on the device and the declared mapper can then accordingly substitute the given map clause with the mapper's contents. However, the mapper is doing a map-type of from for the struct, and an array section. Because the default mapping for a struct of that type has been rewritten, the map-type decay for tofrom -> from = from. However the array section has its own default mapping behavior, which I do not understand from pg. 156 of 5.2:
23 If an array appears as a list item in a map clause, multiple parts of the array have corresponding
24 storage in the device data environment prior to a task encountering the construct associated with
25 the map clause, and the corresponding storage for those parts was created by maps from more
26 than one earlier construct, the behavior is unspecified.
Is this the case for the array section in question?